minpy/dispatch/policy.py

```diff
@@ -38,12 +63,119 @@ def name(self):
+        args : tuple
+            Positional arguments.
+        kwargs : dict
+            Keyword arguments.
+
+        Returns
+        -------
+        Result from appropriate function call.
+        """
+        available = self._available_prims(name, reg, args, kwargs)
+        preference = self.decide(available, args, kwargs)
+        if preference is None:
+            raise PrimitivePolicyError(name, self.name)
+        prim = reg.get(name, preference)
+        _logger.debug('Found primitive "{}" with type {}.'.format(
+            name, prim.typestr))
+        return prim(*args, **kwargs)
+
+
+class AutoBlacklistPolicy(Policy):
+    """Automatically dispatch ops to MXNet impl by provided config.
+
+    Parameters
+    ----------
+    gen_rule : bool
+        If False, use loaded rules to decide. Otherwise, dynamically add new
+        rules and save to rule files.
+    append_rule : bool
+        If True, append new rules to loaded rules. Otherwise, start from
+        scratch.
+    loc : str
+        Path to rule configuration file.
+    """
+
+    def __init__(self, gen_rule=False, append_rule=True, loc=None):
+        self._gen_rule = gen_rule
+        self._rules = Blacklist(loc)
+        if gen_rule:
+            atexit.register(self.save_rules)
+        if gen_rule and not append_rule:
+            self._rules.reset_rules()
+
+    def resolve_call(self, name, reg, args, kwargs):
+        def get_result(impl_type):
+            prim = reg.get(name, impl_type)
+            return prim(*args, **kwargs)
+
+        available = self._available_prims(name, reg, args, kwargs)
+        possible_impl = set(x.type for x in available)
+        if ArrayType.MXNET in possible_impl and self._rules.allow(
+                name, ArrayType.MXNET, args, kwargs):
+            if self._gen_rule:
+                try:
+                    _logger.debug('Try primitive {} with MXNet '
+                                  'implementation.'.format(name))
+                    return get_result(ArrayType.MXNET)
+                except Exception as err:
+                    if ArrayType.NUMPY in possible_impl:
+                        _logger.info('Error occurs. Try primitive {} with '
+                                     'NumPy implementation'.format(name))
+                        self._rules.add(name, ArrayType.MXNET, args, kwargs)
+                        return get_result(ArrayType.NUMPY)
+                    else:
+                        raise err
+            else:
+                _logger.debug('Execute primitive {} with '
+                              'MXNet implementation'.format(name))
+                return get_result(ArrayType.MXNET)
+        elif ArrayType.NUMPY in possible_impl:
+            _logger.debug('Execute primitive {} with '
+                          'NumPy implementation'.format(name))
+            return get_result(ArrayType.NUMPY)
+        else:
+            raise PrimitivePolicyError(name, self.name)
```

Comment:
1. There are some impls only exist in MXNet, such as onehot encoding (I added a simple onehot encoding in NumPy impl last week though).
2. `_available_prims` also checks if gradability. Need to confirm impls have corresponding gradient implementation.

`gen_rule` is redundant. I left it here mostly for testing the coverage of the giving .conf.